packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-engine.ts

```diff
@@ -0,0 +1,888 @@
+
+/**
+ * Creation properties for a `CodePipelineEngine`
+ */
+export interface CodePipelineEngineProps {
+  /**
+   * The name of the CodePipeline pipeline
+   *
+   * @default - Automatically generated
+   */
+  readonly pipelineName?: string;
+
+  /**
+   * Create KMS keys for the artifact buckets, allowing cross-account deployments
+   *
+   * The artifact buckets have to be encrypted to support deploying CDK apps to
+   * another account, so if you want to do that or want to have your artifact
+   * buckets encrypted, be sure to set this value to `true`.
+   *
+   * Be aware there is a cost associated with maintaining the KMS keys.
+   *
+   * @default false
+   */
+  readonly crossAccountKeys?: boolean;
+
+  /**
+   * CDK CLI version to use in self-mutation and asset publishing steps
+   *
+   * If you want to lock the CDK CLI version used in the pipeline, by steps
+   * that are automatically generated for you, specify the version here.
+   *
+   * You should not typically need to specify this value.
+   *
+   * @default - Latest version
+   */
+  readonly cdkCliVersion?: string;
+
+  /**
+   * Whether the pipeline will update itself
+   *
+   * This needs to be set to `true` to allow the pipeline to reconfigure
+   * itself when assets or stages are being added to it, and `true` is the
+   * recommended setting.
+   *
+   * You can temporarily set this to `false` while you are iterating
+   * on the pipeline itself and prefer to deploy changes using `cdk deploy`.
+   *
+   * @default true
+   */
+  readonly selfMutation?: boolean;
+
+  /**
+   * Set if the pipeline itself builds Docker container assets
+   *
+   * NOTE: this only applies to Docker assets used for the Pipeline
+   * or the Pipeline stack itself. It does not apply for Docker assets
+   * used in the Stages and Stacks that are *deployed* by this pipeline.
+   *
+   * Configures privileged mode for the self-mutation CodeBuild action.
+   *
+   * @default false
+   */
+  readonly pipelineUsesDockerAssets?: boolean;
+
+  /**
+   * Customize the CodeBuild projects created for this pipeline
+   *
+   * @default - All projects run non-privileged build, SMALL instance, LinuxBuildImage.STANDARD_5_0
+   */
+  readonly codeBuildDefaults?: CodeBuildDefaults;
+
+  /**
+   * Additional customizations to apply to the asset publishing CodeBuild projects
+   *
+   * @default - Only `codeBuildProjectDefaults` are applied
+   */
+  readonly assetPublishingCodeBuildDefaults?: CodeBuildDefaults;
+
+  /**
+   * Additional customizations to apply to the self mutation CodeBuild projects
+   *
+   * @default - Only `codeBuildProjectDefaults` are applied
+   */
+  readonly selfMutationCodeBuildDefaults?: CodeBuildDefaults;
+
+  /**
+   * Whether this pipeline creates one asset upload action per asset type or one asset upload per asset
+   *
+   * @default false
+   */
+  readonly singlePublisherPerAssetType?: boolean;
```

Comment:
Consider renaming to reflect user intent (e.g. concurrency) instead of mechanism.

Reply:
In the past I've heard you advocate for the opposite, "just describe mechanism and let user figure out why they would need it".

Is this a reversal of the previous guidance?

Reply:
It depends on the level of the layer. In lower level APIs, I am usually leaning towards mechanism because the API is closer to the actual behavior, in high-level APIs such as this one, intent-based APIs are less leaky.

I try to put myself in the shoes of the user and think about where they are coming from. In this case I think "single publisher per asset type" doesn't really mean a lot to users of the `CodePipelineEngine`.